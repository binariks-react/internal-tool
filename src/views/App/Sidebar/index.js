import React, { useState } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const getColor = (color, props) => (
  props ? props.theme.colors[color] : ({ theme: { colors } }) => colors[color]
);

const getActiveColor = (color, props) => (
  props ? props.theme.colors.activeColors[color] : ({ theme: { colors: { activeColors } } }) => activeColors[color]
);

const getHoverColors = (color, props) => (
  props ? props.theme.colors.activeColors[color] : ({ theme: { colors: { hoverColors } } }) => hoverColors[color]
);

const Wrapper = styled.div`
  width: 300px;
  display: block;
  height: 100vh;
  border-right: 1px solid ${getColor('lightGrey')};
`;

const Header = styled.div`
  width: 100%;
  height: 70px;
  border-bottom: 1px solid ${getColor('lightGrey')};
  text-align: center;
  line-height: 70px;
  color: ${getColor('text')};
`;

const Panel = styled.div`
  display: flex;
  align-items: center;
  width: 100%;
  height: 60px;
  border-left: ${({ selected, ...props }) => (
    selected ? `5px solid ${getColor('primary', props)}` : '5px solid transparent'
  )};
  border-bottom: 1px solid ${getColor('lightGrey')};
  padding-left: 30px;
  line-height: 60px;
  background-color: ${({ selected, ...props }) => selected ? getActiveColor('lightGrey', props) : getColor('white', props)};
  color: ${({ selected, ...props }) => selected ? getColor('primary', props) : getColor('text', props)};
  
  &:hover {
    background-color: ${getHoverColors('lightGrey')};
  }
`;

const Icon = styled.i`
  margin-right: 10px;
`;


const Sidebar = ({ titles }) => {
  const [selectedKey, setSelectedKey] = useState(titles.length && titles[0].title.replace(/\s/g, ''));

  return (
    <Wrapper className="wrapper">
      <Header>Dashboards</Header>
      {
        !!titles.length &&
        titles.map(({ title, icon }) => {
          const key = title.replace(/\s/g, '');
          return (
            <Panel
              key={key}
              selected={selectedKey === key}
              onClick={() => setSelectedKey(key)}
            >
              <Icon className="material-icons">
                { icon }
              </Icon>
              <span>{title}</span>
            </Panel>
          );
        })
      }
    </Wrapper>
  );
};

Sidebar.defaultProps = {
  titles: [],
};

Sidebar.propTypes = {
  titles: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.string.isRequired,
    icon: PropTypes.string.isRequired,
  })),
};

export default Sidebar;
