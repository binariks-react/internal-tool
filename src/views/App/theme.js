export const getColor = (color, props) => (
  props ? props.theme.colors[color] : ({ theme }) => theme.colors[color]
);

export const getActiveColor = (color, props) => (
  props ? props.theme.colors.activeColors[color] : ({ theme }) => theme.colors.activeColors[color]
);

export const getHoverColor = (color, props) => (
  props ? props.theme.colors.hoverColors[color] : ({ theme }) => theme.colors.hoverColors[color]
);

export const getHoverShadowColor = (color, props) => (
  props ? props.theme.shadows.hover[color] : ({ theme }) => theme.shadows.hover[color]
);

export const getFocusShadowColor = (color, props) => (
  props ? props.theme.shadows.focus[color] : ({ theme }) => theme.shadows.focus[color]
);

export default {
  colors: {
    text: '#5a6169',
    lightGrey: '#e9ecef',
    darkGrey: '#becad6',
    border: '#e1e5eb',
    primary: '#007bff',
    secondary: '#5a6169',
    success: '#17c671',
    danger: '#c4183c',
    warning: '#ffb400',
    info: '#00b8d8',
    dark: '#212529',
    white: '#ffffff',
    royalBlue: '#674eec',
    java: '#1adba2',
    salmon: '#ff4169',
    hoverColors: {
      lightGrey: '#e9ecef',
      primary: '#006fe6',
      secondary: '#4e545b',
      success: '#14af64',
      danger: '#ad1535',
      warning: '#e6a200',
      info: '#00a2bf',
      dark: '#16181b',
      white: '#ffffff',
      royalBlue: '#5337ea',
      java: '#17c491',
      salmon: '#ff2855',
    },
    activeColors: {
      primary: '#0062cc',
      secondary: '#42484e',
      lightGrey: '#FBFBFB',
      success: '#129857',
      danger: '#97122e',
      warning: '#cc9000',
      info: '#008da5',
      dark: '#0a0c0d',
      white: '#007bff',
      royalBlue: '#3f20e7',
      java: '#15ad80',
      salmon: '#ff0e41',
    },
  },
  shadows: {
    hover: {
      primary: '0 5px 15px rgba(0,0,0,.05), 0 4px 10px rgba(0,123,255,.25)',
      secondary: '0 5px 15px rgba(0,0,0,.05), 0 4px 10px rgba(90,97,105,.25)',
      success: '0 5px 15px rgba(0,0,0,.05), 0 4px 10px rgba(23,198,113,.25)',
      danger: '0 5px 15px rgba(0,0,0,.05), 0 4px 10px rgba(196,24,60,.25)',
      warning: '0 5px 15px rgba(0,0,0,.05), 0 4px 10px rgba(255,180,0,.25)',
      info: '0 5px 15px rgba(0,0,0,.05), 0 4px 10px rgba(0,184,216,.25)',
      dark: '0 5px 15px rgba(0,0,0,.05), 0 4px 10px rgba(33,37,41,.25)',
      white: '0 0.125rem 0.625rem rgba(129,142,163,.2), 0 0.0625rem 0.125rem rgba(129,142,164,.3)',
      royalBlue: '0 5px 15px rgba(0,0,0,.05), 0 4px 10px rgba(103,78,236,.25)',
      java: '0 5px 15px rgba(0,0,0,.05), 0 4px 10px rgba(26,219,162,.25)',
      salmon: '0 5px 15px rgba(0,0,0,.05), 0 4px 10px rgba(255,65,105,.25)',
      inset: 'inset 0 3px 5px rgba(0,0,0,.125)!important',
    },
    focus: {
      primary: '0 0 0 3px rgba(0,123,255,.15),0 3px 15px rgba(0,123,255,.2),0 2px 5px rgba(0,0,0,.1)',
      secondary: '0 0 0 3px rgba(90,97,105,.15),0 3px 15px rgba(90,97,105,.2),0 2px 5px rgba(0,0,0,.1)',
      success: '0 0 0 3px rgba(23,198,113,.15),0 3px 15px rgba(23,198,113,.2),0 2px 5px rgba(0,0,0,.1)',
      danger: '0 0 0 3px rgba(196,24,60,.15),0 3px 15px rgba(196,24,60,.2),0 2px 5px rgba(0,0,0,.1)',
      warning: '0 0 0 3px rgba(255,180,0,.15),0 3px 15px rgba(255,180,0,.2),0 2px 5px rgba(0,0,0,.1)',
      info: '0 0 0 3px rgba(0,184,216,.15),0 3px 15px rgba(0,184,216,.2),0 2px 5px rgba(0,0,0,.1)',
      dark: '0 0 0 3px rgba(33,37,41,.15),0 3px 15px rgba(33,37,41,.2),0 2px 5px rgba(0,0,0,.1)',
      white: '0 0 0 3px rgba(251,251,251,.15),0 3px 15px rgba(251,251,251,.2),0 2px 5px rgba(0,0,0,.1)',
      royalBlue: '0 0 0 3px rgba(103,78,236,.15),0 3px 15px rgba(103,78,236,.2),0 2px 5px rgba(0,0,0,.1)',
      java: '0 0 0 3px rgba(26,219,162,.15),0 3px 15px rgba(26,219,162,.2),0 2px 5px rgba(0,0,0,.1)',
      salmon: '0 0 0 3px rgba(255,65,105,.15),0 3px 15px rgba(255,65,105,.2),0 2px 5px rgba(0,0,0,.1)',
      switch: '0 3px 8px 1px rgba(0,0,0,.2)',
    },
  },
};
