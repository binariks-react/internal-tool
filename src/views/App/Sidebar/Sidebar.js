import React, { useState } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { getColor, getActiveColor, getHoverColor } from 'views/App/theme';

const Wrapper = styled.div`
  position: fixed;
  width: 300px;
  min-width: 300px;
  height: 100vh;
  border-right: 1px solid ${getColor('lightGrey')};
  box-shadow: -15px 0 20px ${getColor('secondary')};
`;

const Header = styled.div`
  width: 100%;
  height: 70px;
  border-bottom: 1px solid ${getColor('lightGrey')};
  text-align: center;
  line-height: 70px;
  cursor: pointer;
  color: ${getColor('text')};
`;

const Panel = styled.div.attrs({ 'data-test': 'panel' })`
  display: flex;
  align-items: center;
  width: 100%;
  height: 60px;
  border-left: ${({ selected, ...props }) => (
    selected ? `5px solid ${getColor('primary', props)}` : '5px solid transparent'
  )};
  border-bottom: 1px solid ${getColor('lightGrey')};
  padding-left: 30px;
  line-height: 60px;
  cursor: pointer;
  background-color: ${({ selected }) => selected ? getActiveColor('lightGrey') : getColor('white')};
  color: ${({ selected }) => selected ? getColor('primary') : getColor('text')};
  transition: all .3s;
  
  &:hover {
    background-color: ${getHoverColor('lightGrey')};
  }
`;

const Icon = styled.i`
  margin-right: 10px;
`;


const Sidebar = ({ titles, onClick, preventDefault, stopPropagation }) => {
  const [selectedKey, setSelectedKey] = useState(titles.length && titles[0].title.replace(/\s/g, ''));

  const handleClick = key => (e) => {
    if (preventDefault) e.preventDefault();
    if (stopPropagation) e.stopPropagation();
    setSelectedKey(key);
    onClick(e);
  };

  return (
    <Wrapper className="wrapper">
      <Header>Dashboards</Header>
      {
        titles.map(({ title, icon }) => {
          const key = title.replace(/\s/g, '');
          return (
            <Panel
              key={key}
              selected={selectedKey === key}
              onClick={handleClick(key)}
            >
              <Icon className="material-icons">
                { icon }
              </Icon>
              <span>{title}</span>
            </Panel>
          );
        })
      }
    </Wrapper>
  );
};

Sidebar.defaultProps = {
  titles: [],
  onClick: () => {},
  preventDefault: true,
  stopPropagation: true,
};

Sidebar.propTypes = {
  titles: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.string.isRequired,
    icon: PropTypes.string.isRequired,
  })),
  onClick: PropTypes.func,
  preventDefault: PropTypes.bool,
  stopPropagation: PropTypes.bool,
};

export default Sidebar;
