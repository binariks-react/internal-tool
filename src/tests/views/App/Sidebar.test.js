import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render, mount } from 'enzyme';
import theme from 'views/App/theme';
import Sidebar from 'views/App/Sidebar/Sidebar';
import 'jest-styled-components';

describe('Sidebar', () => {
  const preventDefault = jest.fn();
  const stopPropagation = jest.fn();
  const onClick = jest.fn();

  beforeEach(() => {
    preventDefault.mockClear();
    stopPropagation.mockClear();
    onClick.mockClear();
  });

  const titles = [
    {
      title: 'Blog Dashboard',
      icon: 'edit',
    },
    {
      title: 'Blog Posts',
      icon: 'vertical_split',
    },
    {
      title: 'Add New Post',
      icon: 'note_add',
    },
  ];

  it('without props expected not to render any panels', () => {
    const sidebar = render(
      <ThemeProvider theme={theme}>
        <Sidebar />
      </ThemeProvider>
    );

    const panels = sidebar.find('[data-test="panel"]');
    expect(panels.length).toBe(0);
    expect(sidebar).toMatchSnapshot();
  });

  describe('with "titles" prop', () => {
    it('expected to render those panels', () => {
      const sidebar = render(
        <ThemeProvider theme={theme}>
          <Sidebar titles={titles} />
        </ThemeProvider>
      );

      const panels = sidebar.find('[data-test="panel"]');
      expect(panels.length).toBe(titles.length);
      expect(sidebar).toMatchSnapshot();
    });

    it('expected first panel to be selected (have "selected" prop)', () => {
      const sidebar = mount(
        <ThemeProvider theme={theme}>
          <Sidebar titles={titles} />
        </ThemeProvider>
      );

      const panel = sidebar.find('[data-test="panel"]').first();
      expect(panel.prop('selected')).toBeDefined();
      expect(panel.prop('selected')).toBe(true);
    });

    it('onClick prop', () => {
      const sidebar = mount(
        <ThemeProvider theme={theme}>
          <Sidebar titles={titles} onClick={onClick} />
        </ThemeProvider>
      );

      const panel = sidebar.find('[data-test="panel"]').at(1);

      panel.simulate('click', { preventDefault, stopPropagation });
      expect(onClick).toHaveBeenCalledTimes(1);
      expect(preventDefault).toHaveBeenCalledTimes(1);
      expect(stopPropagation).toHaveBeenCalledTimes(1);
    });

    it('onClick defaultProps', () => {
      expect(Sidebar.defaultProps.onClick).toBeDefined();
      expect(Sidebar.defaultProps.onClick()).toBeUndefined();
    });

    it('onClick preventDefault = false', () => {
      const sidebar = mount(
        <ThemeProvider theme={theme}>
          <Sidebar titles={titles} onClick={onClick} preventDefault={false} />
        </ThemeProvider>
      );

      const panel = sidebar.find('[data-test="panel"]').at(1);

      panel.simulate('click', { preventDefault, stopPropagation });
      expect(onClick).toHaveBeenCalledTimes(1);
      expect(preventDefault).toHaveBeenCalledTimes(0);
      expect(stopPropagation).toHaveBeenCalledTimes(1);
    });

    it('onClick stopPropagation = false', () => {
      const sidebar = mount(
        <ThemeProvider theme={theme}>
          <Sidebar titles={titles} onClick={onClick} stopPropagation={false} />
        </ThemeProvider>
      );

      const panel = sidebar.find('[data-test="panel"]').at(1);

      panel.simulate('click', { preventDefault, stopPropagation });
      expect(onClick).toHaveBeenCalledTimes(1);
      expect(preventDefault).toHaveBeenCalledTimes(1);
      expect(stopPropagation).toHaveBeenCalledTimes(0);
    });

    it('while clicked on another panel, expected "selected" prop change from false to true', () => {
      const sidebar = mount(
        <ThemeProvider theme={theme}>
          <Sidebar titles={titles} onClick={onClick} />
        </ThemeProvider>
      );

      const selectPanel = index => sidebar.find('[data-test="panel"]').at(index);

      expect(selectPanel(0).prop('selected')).toBeDefined();
      expect(selectPanel(1).prop('selected')).toBeDefined();

      expect(selectPanel(0).prop('selected')).toBe(true);
      expect(selectPanel(1).prop('selected')).toBe(false);

      selectPanel(1).simulate('click');

      sidebar.update();

      expect(selectPanel(0).prop('selected')).toBe(false);
      expect(selectPanel(1).prop('selected')).toBe(true);
    });
  });
});
