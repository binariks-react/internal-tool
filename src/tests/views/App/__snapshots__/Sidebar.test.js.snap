// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar with "titles" prop expected to render those panels 1`] = `
.c0 {
  position: fixed;
  width: 300px;
  min-width: 300px;
  height: 100vh;
  border-right: 1px solid #e9ecef;
  box-shadow: -15px 0 20px #5a6169;
}

.c1 {
  width: 100%;
  height: 70px;
  border-bottom: 1px solid #e9ecef;
  text-align: center;
  line-height: 70px;
  cursor: pointer;
  color: #5a6169;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 60px;
  border-left: 5px solid #007bff;
  border-bottom: 1px solid #e9ecef;
  padding-left: 30px;
  line-height: 60px;
  cursor: pointer;
  background-color: #FBFBFB;
  color: #007bff;
  -webkit-transition: all .3s;
  transition: all .3s;
}

.c2:hover {
  background-color: #e9ecef;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 60px;
  border-left: 5px solid transparent;
  border-bottom: 1px solid #e9ecef;
  padding-left: 30px;
  line-height: 60px;
  cursor: pointer;
  background-color: #ffffff;
  color: #5a6169;
  -webkit-transition: all .3s;
  transition: all .3s;
}

.c4:hover {
  background-color: #e9ecef;
}

.c3 {
  margin-right: 10px;
}

<div
  class="wrapper c0"
>
  <div
    class="c1"
  >
    Dashboards
  </div>
  <div
    class="c2"
    data-test="panel"
    selected=""
  >
    <i
      class="material-icons c3"
    >
      edit
    </i>
    <span>
      Blog Dashboard
    </span>
  </div>
  <div
    class="c4"
    data-test="panel"
  >
    <i
      class="material-icons c3"
    >
      vertical_split
    </i>
    <span>
      Blog Posts
    </span>
  </div>
  <div
    class="c4"
    data-test="panel"
  >
    <i
      class="material-icons c3"
    >
      note_add
    </i>
    <span>
      Add New Post
    </span>
  </div>
</div>
`;

exports[`Sidebar without props expected not to render any panels 1`] = `
.c0 {
  position: fixed;
  width: 300px;
  min-width: 300px;
  height: 100vh;
  border-right: 1px solid #e9ecef;
  box-shadow: -15px 0 20px #5a6169;
}

.c1 {
  width: 100%;
  height: 70px;
  border-bottom: 1px solid #e9ecef;
  text-align: center;
  line-height: 70px;
  cursor: pointer;
  color: #5a6169;
}

<div
  class="wrapper c0"
>
  <div
    class="c1"
  >
    Dashboards
  </div>
</div>
`;
